File: fileUtils.js
const fs = require('fs').promises;
const path = require('path');
const { config } = require('./config');
const { logger } = require('./logger');
const { BotError } = require('./errors');

class FileUtils {
    /**
     * Save torrent file to disk
     * @param {string} kzId - Kinozal torrent ID
     * @param {Buffer} data - Torrent file data
     * @returns {Promise<string>} Path to saved torrent file
     */
    async saveTorrentFile(kzId, data) {
        try {
            const torrentPath = path.join(config.folders.torrents, `${kzId}.torrent`);
            await fs.writeFile(torrentPath, data);
            logger.debug(`Torrent file saved: ${torrentPath}`);
            return torrentPath;
        } catch (error) {
            throw new BotError(
                'Failed to save torrent file',
                'FILE_SYSTEM_ERROR',
                { kzId, originalError: error.message }
            );
        }
    }

    /**
     * Clean up torrent file after it's added to Transmission
     * @param {string} torrentPath - Path to torrent file
     */
    async cleanupTorrentFile(torrentPath) {
        try {
            await fs.unlink(torrentPath);
            logger.debug(`Torrent file cleaned up: ${torrentPath}`);
        } catch (error) {
            // Log but don't throw - cleanup failure shouldn't stop the process
            logger.warn('Failed to cleanup torrent file', {
                torrentPath,
                error: error.message
            });
        }
    }



    /**
     * Get file size
     * @param {string} filePath - File path
     * @returns {Promise<number>} File size in bytes
     */
    async getFileSize(filePath) {
        try {
            const stats = await fs.stat(filePath);
            return stats.size;
        } catch (error) {
            throw new BotError(
                'Failed to get file size',
                'FILE_SYSTEM_ERROR',
                { path: filePath, originalError: error.message }
            );
        }
    }
}

// Export singleton instance
module.exports = {
    fileUtils: new FileUtils()
};

File: src/fileutils/fileutils.go
package fileutils

import (
	"fmt"
	"os"
	"path/filepath"

	"kinozal-bot/errors"
	"kinozal-bot/logger"
)

// SaveTorrentFile saves the torrent file to disk
func SaveTorrentFile(kzID string, data []byte) (string, error) {
	// Определяем путь к файлу
	torrentDir := "torrents"
	torrentPath := filepath.Join(torrentDir, fmt.Sprintf("%s.torrent", kzID))

	// Проверяем, существует ли директория
	err := os.MkdirAll(torrentDir, os.ModePerm)
	if err != nil {
		return "", errors.NewBotError(
			"Failed to create directory for torrent files",
			"FILE_SYSTEM_ERROR",
			map[string]interface{}{
				"directory":      torrentDir,
				"original_error": err.Error(),
			},
		)
	}

	// Сохраняем файл
	err = os.WriteFile(torrentPath, data, 0644)
	if err != nil {
		return "", errors.NewBotError(
			"Failed to save torrent file",
			"FILE_SYSTEM_ERROR",
			map[string]interface{}{
				"kzID":           kzID,
				"original_error": err.Error(),
			},
		)
	}

	logger.Debug("Torrent file saved", map[string]interface{}{
		"path": torrentPath,
	})
	return torrentPath, nil
}

// CleanupTorrentFile removes the torrent file after use
func CleanupTorrentFile(torrentPath string) error {
	err := os.Remove(torrentPath)
	if err != nil {
		logger.Warn("Failed to cleanup torrent file", map[string]interface{}{
			"path":           torrentPath,
			"original_error": err.Error(),
		})
		return err
	}

	logger.Debug("Torrent file cleaned up", map[string]interface{}{
		"path": torrentPath,
	})
	return nil
}

// GetFileSize returns the size of a file in bytes
func GetFileSize(filePath string) (int64, error) {
	fileInfo, err := os.Stat(filePath)
	if err != nil {
		return 0, errors.NewBotError(
			"Failed to get file size",
			"FILE_SYSTEM_ERROR",
			map[string]interface{}{
				"path":           filePath,
				"original_error": err.Error(),
			},
		)
	}
	return fileInfo.Size(), nil
}